twenty-first/src/util_types/mmr/mmr_trait.rs

```diff
@@ -39,42 +39,129 @@ impl LeafMutation {
     }
 }
 
+/// A “Merkle Mountain Range” (MMR) is a data structure for storing lists of
+/// hashes. Similar to [Merkle trees](crate::prelude::MerkleTree), they enable
+/// efficient set inclusion proofs. Unlike Merkle trees, the number of leafs of
+/// an MMR is not required to be a power of 2.
+///
+/// One valid perspective of this data structure is to see it as a list of
+/// Merkle trees. Each of the Merkle tree's roots is one peak of the MMR.
+/// In order to keep the number of peaks to a minimum, any two neighboring
+/// Merkle trees of the same height are merged, which results in a single tree
+/// that has one additional “tier”.
+///
+/// # Example
+///
+/// The following is a Merkle Mountain Range, annotated with both node and leaf
+/// indices:
+///
+/// ```markdown
+///         ──── 14 ───                           ╮
+///        ╱           ╲                          │
+///       6             13            21          │
+///      ╱  ╲          ╱  ╲          ╱  ╲         ├╴ node indices
+///     ╱    ╲        ╱    ╲        ╱    ╲        │
+///    2      5      9      12     17     20      │
+///   ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲      │
+///  0   1  3   4  7   8  10 11  15 16  18 19 22  ╯
+///
+///  0   1  2   3  4   5  6   7  8   9  10 11 12  ←── leaf indices
```

Comment:
The node indices here are off-by-one compared to the convention we're using. The leaf indices are correct.